### PR TITLE
fix: validation for arraykeys in suite for it.each cases

### LINKED ISSUE #8189: 3.1.0+ expands strings matching $\d+, inside it.each items unexpectedly
### Describe the bug

Given the following test:
```
const dollars = (amount: number) => {
    return new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(amount)
}

it.each([
  [0, "$0.00"],
  [34, "$34.00"],
  [3434, "$3,434.00"],
  [343434, "$343,434.00"],
])("handles whole numbers: %s as %s", (input, expected) => {
  expect(dollars(input)).toBe(expected);
});
```

in vitest 3.0.9, you get the expected output:
```
 ✓ src/dollars.test.ts (5 tests) 20ms
   ✓ handles whole numbers: 0 as $0.00
   ✓ handles whole numbers: 34 as $34.00
   ✓ handles whole numbers: 3434 as $3,434.00
   ✓ handles whole numbers: 343434 as $343,434.00

 Test Files  1 passed (1)
      Tests  5 passed (5)
```

However in 3.1.0-beta.1 and newer, you get the following output:
```
 ✓ src/dollars.test.ts (4 tests) 20ms
   ✓ handles whole numbers: 0 as $0.00 18ms
   ✓ handles whole numbers: 34 as $34.00 0ms
   ✓ handles whole numbers: 3434 as undefined,434.00 0ms
   ✓ handles whole numbers: 343434 as undefined,434.00 0ms

 Test Files  1 passed (1)
      Tests  4 passed (4)
```

I believe the highlighted code in this commit is what is causing the issue.
https://github.com/vitest-dev/vitest/pull/7522/files#diff-52be9c47a2c7954c486556e66d29729f4259c31adbe08801a36cfde26c245307R816-R821

As a user of vitest, I expect variable expansion to only work from the template string, not from within item strings. However the `.replace` to replace object/array keys happens on the final formatted string, after `%s` and similar replacements have been expanded. This results in strings inside the it.each items being unexpectedly expanded as well. To demonstrate that this is what is happening, consider the following:
```
const testrow: [number, string] = [343434, "$343,434.00"];
testrow["343"] = "foobar";

it.each([testrow])("handles whole numbers: %s as %s", (input, expected) => {
  expect(dollars(input)).toBe(expected);
});
```
This outputs `✓ handles whole numbers: 343434 as 'foobar',434.00 18ms`, demonstrating that it is interpreting the $343 as an arrayKey. 

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-g7rfid5c?file=test%2Fbasic.test.ts

### System Info

```shell
System:
    OS: Linux 6.6 Ubuntu 24.04.2 LTS 24.04.2 LTS (Noble Numbat)
    CPU: (8) x64 AMD EPYC 7B12
    Memory: 28.36 GB / 31.35 GB
    Container: Yes
    Shell: 5.2.21 - /bin/bash
  Binaries:
    Node: 20.18.3 - ~/.volta/tools/image/node/20.18.3/bin/node
    Yarn: 1.22.22 - ~/.volta/bin/yarn
    npm: 10.8.2 - ~/.volta/tools/image/node/20.18.3/bin/npm
    pnpm: 10.12.1 - ~/.volta/bin/pnpm
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by DevJoaoLopes**
### Description

Change validation within `suite.ts` to fix switching issue in scenarios using` it.each`.

Used an early return to prevent tests from falling into an incorrect flow.

fixes #8189 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by DevJoaoLopes** [comment]
@hi-ogawa As I had tagged you in the[ issue debate](https://github.com/vitest-dev/vitest/issues/8189#issuecomment-2989742841), can you give me some support on how we will proceed with the adjustment?

**Review by hi-ogawa — COMMENTED** [review]
Can you also add test case from https://github.com/vitest-dev/vitest/issues/8189#issuecomment-2989378534?

**Comment by DevJoaoLopes** [comment]
> Can you also add test case from [#8189 (comment)](https://github.com/vitest-dev/vitest/issues/8189#issuecomment-2989378534)?

Done @hi-ogawa

**Comment by hi-ogawa** [comment]
Thanks for the effort, however this only changes the order of expansion from `% -> $` to `$ -> %` and that's not a proper solution as I wrote in https://github.com/vitest-dev/vitest/issues/8189#issuecomment-3088609224.  For example, this previously passed but now fails:

```js
test.for([{ a: '%o' }])('$a', (_, ctx) => {
  expect(ctx.task.name).toBe("'%o'")
// Expected: "'%o'"
// Received: "'{ a: '%o' }'"
});
```

Also please add an test case in a way it satisfies the 2nd checkbox in PR description. You only added only examples and they don't fail on current main.
> [ ] Ideally, include a test that fails without this PR but passes with it.

There are similar test cases already in https://github.com/vitest-dev/vitest/blob/c1ac15c6b22903cf17b5b105dffd0b8aabf083c3/test/reporters/tests/default.test.ts#L202

**Comment by DevJoaoLopes** [comment]
> Thanks for the effort, however this only changes the order of expansion from `% -> $` to `$ -> %` and that's not a proper solution as I wrote in [#8189 (comment)](https://github.com/vitest-dev/vitest/issues/8189#issuecomment-3088609224). For example, this previously passed but now fails:
> 
> ```js
> test.for([{ a: '%o' }])('$a', (_, ctx) => {
>   expect(ctx.task.name).toBe("'%o'")
> // Expected: "'%o'"
> // Received: "'{ a: '%o' }'"
> });
> ```
> 
> Also please add an test case in a way it satisfies the 2nd checkbox in PR description. You only added only examples and they don't fail on current main.
> 
> > [ ] Ideally, include a test that fails without this PR but passes with it.
> 
> There are similar test cases already in
> 
> https://github.com/vitest-dev/vitest/blob/c1ac15c6b22903cf17b5b105dffd0b8aabf083c3/test/reporters/tests/default.test.ts#L202

@hi-ogawa It really makes sense, maybe it's more complex than we imagined 😄 . I implemented an idea, can I send the code here for you to take a look at or do you prefer that I commit it?

**Comment by hi-ogawa** [comment]
Yes, this is tricky and I'm not sure the fix is that simple either. What I have in mind would roughly look like a following process:

```
"$a %o $b %s"

  ⇓ (replace $ with placeholder)

"__tmp1__ %o __tmp2__ %s"

  ⇓ (apply % formatting)

"__tmp1__ xxx __tmp2__ yyy"

---

"$a"  "$b"

  ⇓    ⇓ (apply $ formatting individually)

"www" "zzz"

---

"__tmp1__ xxx __tmp2__ yyy"

  ⇓ (restore $ placeholder)

"www xxx zzz yyy"
```

**Comment by DevJoaoLopes** [comment]
my idea would be the following, _looking at the code_:

```ts
function formatTitle(template: string, items: any[], idx: number) {
  // Step 1: Replace escaped percent signs
  template = template.replace(/%%/g, '__vitest_escaped_%__')

  // Step 2: Replace % expansions
  if (template.includes('%#') || template.includes('%$')) {
    template = template
      .replace(/%#/g, `${idx}`)
      .replace(/%\$/g, `${idx + 1}`)
  }

  // Step 3: Replace $ expansions
  // We want to expand $vars in the template, but only outside of %f or other %... expansions
  // We'll split the template into segments: those inside %f (or other %X) and those outside
  // For simplicity, let's process $expansions everywhere, but only if not inside a %... sequence

  // Find all %f, %s, etc. and replace them with placeholders
  const percentRegex = /%[a-z]/gi
  const percentMatches: RegExpMatchArray | null = template.match(percentRegex)
  const percentCount = percentMatches ? percentMatches.length : 0

  // Replace $expansions
  const isObjectItem = isObject(items[0])
  template = template.replace(
    /\$([$\w.]+)/g,
    (_, key: string) => {
      const isArrayKey = /^\d+$/.test(key)
      if (!isObjectItem && !isArrayKey) {
        return `$${key}`
      }
      const arrayElement = isArrayKey ? objectAttr(items, key) : undefined
      const value = isObjectItem ? objectAttr(items[0], key, arrayElement) : arrayElement
      return objDisplay(value, {
        truncate: runner?.config?.chaiConfig?.truncateThreshold,
      }) as unknown as string
    },
  )

  // Step 4: Restore escaped percent signs
  template = template.replace(/__vitest_escaped_%__/g, '%')

  // Step 5: Handle %f special case for -0 and NaN
  if (template.includes('%f')) {
    const placeholders = template.match(/%f/g) || []
    let temporaryTemplate = template
    placeholders.forEach((_, i) => {
      if (isNegativeNaN(items[i]) || Object.is(items[i], -0)) {
        let occurrence = 0
        temporaryTemplate = temporaryTemplate.replace(/%f/g, (match) => {
          occurrence++
          return occurrence === i + 1 ? '-%f' : match
        })
      }
    })
    return format(temporaryTemplate, ...items.slice(0, percentCount))
  }

  return format(template, ...items.slice(0, percentCount))
}
```

@hi-ogawa

**Comment by hi-ogawa** [comment]
Please feel free to update the code and let's see the test case 🙂

**Comment by DevJoaoLopes** [comment]
> Please feel free to update the code and let's see the test case 🙂

I uploaded the code I commented on.

Regarding the reporter tests, I inserted it into the file you mentioned, but I'm not sure if it's the best place. I can change it if I see fit

@hi-ogawa

File: test/core/test/each.test.ts [hunk 1]
```diff
@@ -238,3 +238,13 @@ ${{ val: 1 }} | ${{ val: 2 }}} | ${3}
...
+
+test.each([[343434, '$343,434.00']])('handles whole numbers: %s as %s', (input, expected) => {
+  expect(new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(input)).toBe(expected)
+})
+
+test.each([{ a: '$b', b: 'yay' }])('%o', () => {})
+
+test.for([{ a: '%o' }])('$a', (_, ctx) => {
+  expect(ctx.task.name).toBe('\'%o\'')
+})
```
Comment (hi-ogawa) [inline test/core/test/each.test.ts:250]: You can move this to `fixtures/test-for-title.test.ts` 

https://github.com/vitest-dev/vitest/blob/d19ff43f2a61f7381cbf60c37edc65c2831b77e6/test/reporters/tests/default.test.ts#L202-L207

**Review by hi-ogawa — COMMENTED** [review]
It looks like this only changed the order from `format -> objDisplay` to `objDisplay -> format`. Though it looks like tests are passing, can you explain how this approach avoids the issue pointed out in https://github.com/vitest-dev/vitest/pull/8192#issuecomment-3195290268?

**Review by hi-ogawa — CHANGES_REQUESTED** [review]
I think `percentCount` slice doesn't save an edge case like:

```js
test.each([
  { a: "%o" }
])('$a %o', () => {
});

// ✓ '{ a: '%o' }' %o 0ms
```

The expected title is

```
//  ✓ '%o' { a: '%o' } 0ms
```
